features/26-mfa/api.md

```diff
@@ -0,0 +1,436 @@
+# Overview
+
+The following factors are supported:
+
+- Time-Based One-Time Password [(TOTP)](https://tools.ietf.org/html/rfc6238)
+- Out-of-band
+  - SMS
+- Recovery code
+- Bearer token
+
+In the future, the below factors will be supported:
+
+- [WebAuthn with CTAP2 or FIDO U2F](https://fidoalliance.org/specifications/)
+
+The developer can configure the enforcement of MFA.
+
+The access token contains how the user was authenticated.
+
+# Configuration
+
+```yaml
+mfa:
+  # 'optional' or 'required'
+  #
+  # 'optional' means the user is free to register a new authenticator.
+  # When a user has at least one active authenticator, the user must login with MFA.
+  # This is like how other online services such as Google and Facebook offer MFA.
+  #
+  # 'required' means the user must login with MFA.
+  # APIs like change password will return MFA required error.
+  # This is like how banking services offer MFA.
+  #
+  # Default is 'optional'
+  enforcement: 'optional'
```

Comment:
You mean within the same app, some users (maybe they are admin) require MFA while some other users (maybe they are casual users) do not? Since we do not have authorization now and the auth gear does not have the concept of admin (we use master key), the application should use the injected `amr` header to perform authorization themselves.

Reply:
@louischan-oursky so do you meant the way it should works is:

1. Developers set the enforcement to optional
2. Developers write the code so that in the `amr` header, if there are no mfa but requires for the users or other situation, the code should return error to client side, and the client side will trigger the 2FA login manually?

If that's what you meant (and what current design can support) I think is okay.

Reply:
Yes.